src/rbus/rbus.c

```diff
@@ -2333,46 +2504,79 @@ static void _subscribe_callback_handler (rbusHandle_t handle, rbusMessage reques
     }
 }
 
-static int _callback_handler(char const* destination, char const* method, rbusMessage request, void* userData, rbusMessage* response, const rtMessageHeader* hdr)
+static int _callback_handler(
+  char const*               destination,
+  char const*               method,
+  rbusMessage               request,
+  void*                     userData,
+  rbusMessage*              response,
+  const rtMessageHeader*    header)
 {
-    rbusHandle_t handle = (rbusHandle_t)userData;
-
-    RBUSLOG_DEBUG("Received callback for [%s]", destination);
+  int ret = 0;
+  rbusHandle_t rbus = (rbusHandle_t) userData;
+  if (rbus->useEventLoop) {
+    struct rbusEventLoopCallContext ctx;
+    ctx.future = rtFuture_Create();
+    ctx.destination = destination;
+    ctx.method = method;
+    ctx.request = request;
+    ctx.user_data = userData;
+    ctx.response = response;
+    ctx.header = header;
+
+    rbusRunnable_t r;
+    r.argp = &ctx;
+    r.exec = &rbusCallbackHandlerAdapter;
+    r.cleanup = NULL;
+    r.next = NULL;
+    r.id = 0;
+
+    rbusRunnableQueue_PushBack(&rbus->eventQueue, r);
+
+    // wait for eventloop to actually process the above request
+    // once it's completed, we can resume
+    rtError err = rtFuture_Wait(ctx.future, 10000); // TODO: where does timeout come from?
```

Comment:
This and other rtFuture_Wait() calls concern me.

If this times out the program is essentially a deadman walking: the eventQueue retains a runnable that refers to an originally stack allocated object `rbusEventLoopCallContext`.

This runs in a separate worker thread, right? The intention here AFAIU is to serialize against other callback handlers and/or provide a backpressure? Ideally these callbacks don't need to be serialized in the worker thread since they will all be processed in-order by the event loop driven consumer, but I can imagine the core might have some implicit assumptions about ordering of other things associated with handling callbacks so I guess this is fine for initial implementation. This could run into performance issues though since this forces the event loop to handle some of the runnables one-by-one.

In either case I would prefer this wait to be indefinite. Any accidental stall in program (blocking syscall that could happen due to RTNL being held by a buggy/slow WLAN driver for example) could lead to an inevitable crash with this timeout logic.

If you insist on keeping the timeout it should literally SIGABRT or something on timeout to make it easier to debug. Debugging dangling pointers is no fun.

Reply:
I agree on TIMEOUT_INFINITE. This is dispatching from within rbus to the application. If the application hangs, the calling thread should also hang

Reply:
a separate checkin to address the timeout will be made